### ISSUE TITLE
Performance regression on upstream kernel 4.20

### ISSUE THREAD

**Issue by sebpuetz** [ROCmSoftwarePlatform/tensorflow-upstream#329]
<em>Please make sure that this is a build/installation issue. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:build_template</em>

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Mint 19.1
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): binary (pip)
- TensorFlow version: 1.12
- Python version: 3.6.7
- Installed using virtualenv? pip? conda?:  pyenv
- Bazel version (if compiling from source): 
- GCC/Compiler version (if compiling from source):
- ROCm/MIOpen version: 2.1.96
- GPU model and memory: Radeon VII 16GB

**Describe the problem**
@sunway513 Continuation from #173 , regression in performance on cnn benchmarks with upstream kernel 4.20 on Linux Mint 19.1

**Provide the exact sequence of commands / steps that you executed before running into the problem**

**Any other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by sunway513** [comment]
Thanks @sebpuetz :-)
As the first step, could you try to purge rock-dkms driver on your installation?
I need to warn that poke around kernel modules can be risky, depends on your local configurations. We'd suggest you properly backup your system before proceeding.
The following step should help reinstall only with the rocm user-bit packages:
```
sudo apt autoremove rock-dkms
sudo apt install rocm-dev  rocm-libs miopen-hip cxlactivitylogger
sudo rebot
```

**Comment by sebpuetz** [comment]
The system shouldn't include anything ROCm related anymore

```
dpkg -l | grep ^ii | awk '{print $2}' | grep roc
amd64-microcode
intel-microcode
libcroco3:amd64
libpostproc54:amd64
libprocps6:amd64
libwebrtc-audio-processing1:amd64
procps
python-setproctitle:amd64
python3-ptyprocess
python3-setproctitle:amd64
```

```
dpkg -l | grep ^ii | awk '{print $2}' | grep mio
librpmio8
```

**Comment by sebpuetz** [comment]
Reinstalled the components through `sudo apt install rocm-dev  rocm-libs miopen-hip cxlactivitylogger` but performance seems to stay the same:

```
TF_ROCM_FUSION_ENABLE=1 python tf_cnn_benchmarks.py --num_gpus=1 --batch_size=128 --model=resnet50

Step	Img/sec	total_loss
1	images/sec: 227.3 +/- 0.0 (jitter = 0.0)	8.221
10	images/sec: 226.6 +/- 0.2 (jitter = 0.4)	8.289
20	images/sec: 226.3 +/- 0.2 (jitter = 0.6)	8.054
30	images/sec: 226.1 +/- 0.2 (jitter = 0.8)	8.317
40	images/sec: 226.1 +/- 0.1 (jitter = 0.8)	8.196
50	images/sec: 226.0 +/- 0.1 (jitter = 0.8)	8.244
60	images/sec: 225.9 +/- 0.1 (jitter = 0.8)	8.193
70	images/sec: 226.0 +/- 0.1 (jitter = 0.7)	8.148
80	images/sec: 226.1 +/- 0.1 (jitter = 0.6)	8.157
90	images/sec: 226.1 +/- 0.1 (jitter = 0.6)	8.208
100	images/sec: 225.9 +/- 0.1 (jitter = 0.7)	8.179
----------------------------------------------------------------
total images/sec: 225.88
----------------------------------------------------------------
```

**Comment by sunway513** [comment]
Thanks @sebpuetz , let me experiment 4.20 kernel a bit more on Ubuntu, will update after I got more clues.

**Comment by sebpuetz** [comment]
I'm only on an upstream kernel because installing `rock-dkms` through the apt repo fails while building the kernel module:
```
Building initial module for 4.15.0-45-generic
ERROR (dkms apport): kernel package linux-headers-4.15.0-45-generic is not supported
Error! Bad return status for module build on kernel: 4.15.0-45-generic (amd64)
Consult /var/lib/dkms/amdgpu/2.1-96/build/make.log for more information.
```
If there is a quick workaround for this, I'd also be glad to hear about that, the attachment includes the build log.
[make.log](https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/files/2890121/make.log)

**Comment by sunway513** [comment]
Hi @sebpuetz , that should be a limitation from Linux Mint, I've seen the same issue there when I trying to install rock-dkms on Linux-mint kernel. 
Need to mention that though Linux Mint is based on Ubuntu, it's not in our official [support list](https://github.com/RadeonOpenCompute/ROCm#supported-operating-systems---new-operating-systems-available).

We are not saying the Linux distros other than the ones listed on the doc won't work for ROCm. However, since those are not in ROCm official support list and not been included in ROCm regular testing cyclse, it's in the YMMV category, unfortunately.
In order to achieve the optimal performance and stability, I'd suggest to stick with the official supported distros:
```
Ubuntu 16.04.x and 18.04.x (Version 16.04.3 and newer or kernels 4.13 and newer)
CentOS 7.4, 7.5, and 7.6 (Using devtoolset-7 runtime support)
RHEL 7.4, 7.5, and 7.6 (Using devtoolset-7 runtime support)
```

BTW, I'm upgrading my dev system to Ubuntu18.04, to work with 4.20 kernel, please stay tuned.

**Comment by sunway513** [comment]
Hi @sebpuetz , as I've synced with our kernel experts, you might want to date the firmware version for Radeon VII with 4.20 kernel, steps as follows:
```
cd ~ && git clone  http://git.kernel.org/pub/scm/linux/kernel/git/dwmw2/linux-firmware.git
cd ~/linux-firmware && sudo cp amdgpu/vega20_*.bin /lib/firmware/amdgpu/
sudo update-initramfs -u
```
Again, please properly backup your system before applying the above steps :-)

**Comment by sebpuetz** [comment]
Rebooting now
```
sudo update-initramfs -u
update-initramfs: Generating /boot/initrd.img-4.20.7-042007-generic
W: Possible missing firmware /lib/firmware/amdgpu/raven2_gpu_info.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/picasso_gpu_info.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/vega12_gpu_info.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/hawaii_k_smc.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/hawaii_smc.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/bonaire_k_smc.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/bonaire_smc.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/mullins_mec.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/mullins_rlc.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/mullins_ce.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/mullins_me.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/mullins_pfp.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/kabini_mec.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/kabini_rlc.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/kabini_ce.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/kabini_me.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/kabini_pfp.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/kaveri_mec2.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/kaveri_mec.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/kaveri_rlc.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/kaveri_ce.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/kaveri_me.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/kaveri_pfp.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/hawaii_mec.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/hawaii_rlc.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/hawaii_ce.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/hawaii_me.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/hawaii_pfp.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/bonaire_mec.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/bonaire_rlc.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/bonaire_ce.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/bonaire_me.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/bonaire_pfp.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/mullins_sdma1.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/mullins_sdma.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/kabini_sdma1.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/kabini_sdma.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/kaveri_sdma1.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/kaveri_sdma.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/hawaii_sdma1.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/hawaii_sdma.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/bonaire_sdma1.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/bonaire_sdma.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/si58_mc.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/hainan_mc.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/oland_mc.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/verde_mc.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/pitcairn_mc.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/tahiti_mc.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/hainan_rlc.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/hainan_ce.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/hainan_me.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/hainan_pfp.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/oland_rlc.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/oland_ce.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/oland_me.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/oland_pfp.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/verde_rlc.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/verde_ce.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/verde_me.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/verde_pfp.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/pitcairn_rlc.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/pitcairn_ce.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/pitcairn_me.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/pitcairn_pfp.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/tahiti_rlc.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/tahiti_ce.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/tahiti_me.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/tahiti_pfp.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/banks_k_2_smc.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/hainan_k_smc.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/hainan_smc.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/oland_k_smc.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/oland_smc.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/verde_k_smc.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/verde_smc.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/pitcairn_k_smc.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/pitcairn_smc.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/tahiti_smc.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/hawaii_mc.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/bonaire_mc.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/vega12_asd.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/vega12_sos.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/raven2_asd.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/picasso_asd.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/raven2_rlc.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/raven2_mec2.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/raven2_mec.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/raven2_me.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/raven2_pfp.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/raven2_ce.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/picasso_rlc.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/picasso_mec2.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/picasso_mec.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/picasso_me.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/picasso_pfp.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/picasso_ce.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/vega12_rlc.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/vega12_mec2.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/vega12_mec.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/vega12_me.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/vega12_pfp.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/vega12_ce.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/raven2_sdma.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/picasso_sdma.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/vega12_sdma1.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/vega12_sdma.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/vega12_uvd.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/mullins_uvd.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/hawaii_uvd.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/kaveri_uvd.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/kabini_uvd.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/bonaire_uvd.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/vega12_vce.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/mullins_vce.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/hawaii_vce.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/kaveri_vce.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/kabini_vce.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/bonaire_vce.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/raven2_vcn.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/picasso_vcn.bin for module amdgpu
W: Possible missing firmware /lib/firmware/amdgpu/vega12_smc.bin for module amdgpu
W: initramfs-tools configuration sets RESUME=UUID=eb946697-2896-48f0-bf88-9d9dfa9a91be
W: but no matching swap device is available.
I: The initramfs will attempt to resume from /dev/sdb6
I: (UUID=6fc3bc62-cdf2-4907-8e9e-0d579b4ff073)
I: Set the RESUME variable to override this.
```

**Comment by sebpuetz** [comment]
Can't see any change in performance

**Comment by sunway513** [comment]
Thanks @sebpuetz  for trying it out.
I believe the low-hanging fruit is to use our main-stream support config - Ubuntu16.04/18.04, with the 4.15 kernel + rock-dkms kernel driver. It's been confirmed that Radeon VII perf should be as expected there. 

BTW, I've got issues boot 4.20 kernel on my local dev node.. Will update here after we have more info to share.

**Comment by sebpuetz** [comment]
> Thanks @sebpuetz for trying it out.
> I believe the low-hanging fruit is to use our main-stream support config - Ubuntu16.04/18.04, with the 4.15 kernel + rock-dkms kernel driver. It's been confirmed that Radeon VII perf should be as expected there.

I'll do that, probably going to install Ubuntu later tonight or tomorrow. I'll keep my Mint installation around so I can check if something will fix the issues for 4.20 on Mint. 
I'll also check with the benchmark scores from the Ubuntu system, that should rule out anything hardware related then

**Comment by sunway513** [comment]
@sebpuetz I'll close this issue for now. Let's track the upstream kernel support on the main ROCm repo.
